### ISSUE TITLE
Thread pool does not create enough workers to handle multiple invocations of calllInThread with blocking functions

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8090]
|<img alt="bkambach's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| bkambach reported|
|-|-|
|Trac ID|trac#8090|
|Type|defect|
|Created|2015-11-06 19:05:07Z|
|Branch|https://github.com/twisted/twisted/tree/threadstarter-8090|

OS: Linux (Fedora) 3.10 kernel
Python: 2.7.10
Twisted: 15.4.0

If I queue multiple jobs via reactor.callInThread() and one of the jobs is blocking, none of the jobs queued after the blocking job are executed.  However, if I manually increase the thread pool size, I can get all of the jobs to execute.

The following example code works in 15.3.0 but not in 15.4.0.  If you un-comment the commented-out code, it will work in 15.4.0:
```
import socket
from twisted.internet import reactor

def blocking_fn(arg):
    port = 2000 + int(arg)
    print "I'm reactor thread %d trying to listen on port %d" % (arg, port)
    ai = socket.getaddrinfo("127.0.0.1", port, socket.AF_INET, socket.SOCK_STREAM)
    assert ai is not None
    sock = socket.socket(ai[0][0], ai[0][1])
    sock.bind(ai[0][4])
    sock.listen(5)
    sock.accept()

def non_blocking_fn(arg):
    print "I'm reactor thread %d" % arg

def main():
    reactor.callInThread(blocking_fn, 1)
    reactor.callInThread(non_blocking_fn, 2)
    reactor.callInThread(non_blocking_fn, 3)
    reactor.callInThread(non_blocking_fn, 4)

    # Enable this code to work around the problem in 15.4.0
    # thread_pool = reactor.getThreadPool()
    # thread_pool.adjustPoolSize(4)
    
    reactor.run()
    
if __name__ == '__main__':
    main()
```

I understand that using raw sockets like this is not recommended with Twisted, but using it is the easiest way to demonstrate the problem I'm seeing.  I suspect this is related to the changes incorporated for [#2673](https://github.com/twisted/twisted/issues/2673).

<details><summary>Searchable metadata</summary>

```
trac-id__8090 8090
type__defect defect
reporter__bkambach bkambach
priority__normal normal
milestone__None None
branch__branches_threadstarter_8090 branches/threadstarter-8090
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1446836707410155 1446836707410155
changetime__1447840260196282 1447840260196282
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Related: [#8043](https://github.com/twisted/twisted/issues/8043)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Related: [#5359](https://github.com/twisted/twisted/issues/5359)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Related: [#4347](https://github.com/twisted/twisted/issues/4347)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Related: [#5298](https://github.com/twisted/twisted/issues/5298)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Thanks for the bug report, bkambach.  As you can see from the above comments, I thought that there was maybe a duplicate of this filed already - and maybe [#8043](https://github.com/twisted/twisted/issues/8043) is it.  But this is described much more clearly :).

The bug here is that initially, no threads are started, because the reactor threadpool isn't running yet. Then, once it starts running, `ThreadPool.start` calls `adjustPoolsize` to start enough threads to fulfill the _minimum_ threadpool size.  However, the only examination of the backlog of existing work in `ThreadPool.start` is this:

```
        if self._team.statistics().backloggedWorkCount:
            self._team.grow(1)
```

in other words, if there is _any_ backlogged work, start only a single worker.  This should probably be more aggressive.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

That said, `twisted.python.threadpool` and, by extension, `reactor.callInThread`, are somewhat crummy APIs for doing what you are trying to do here, which is "coordinate a bunch of long-running tasks".  `callInThread` is generally there for tasks which may block for a while, but will ultimately complete even without coordination with other tasks in the reactor.

There _isn't_ yet a good public API for doing anything else, but you may want to look into the not-yet-public `twisted._threads` module, which is the basis for `twisted.python.threadpool`'s current implementation:

https://twistedmatrix.com/documents/15.4.0/api/twisted._threads.html

If you want to create a long-running thread that you can manage and shut down, you can just call `ThreadWorker().do(lambda: non_blocking_fn(1))` and that will unconditionally start a thread.  You can then `.quit()` that `ThreadWorker` to clean it up.  This is nice because you can separately call `.do(bind())`, `.do(listen())` and `.do(accept())` so that you can easily keep a reference to your FD (and therefore close it) from the main thread.

Alternately, you could just make a native `threading.Thread`, although that only lets you tell it to do one thing.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Finally, there's a workaround for you that will work on current releases: use `reactor.callWhenRunning`, like so:

```
#!py
    @reactor.callWhenRunning                                                                                                                                                         
    def startThreads():                                                                                                                                                              
        reactor.callInThread(blocking_fn, 1)                                                                                                                                         
        reactor.callInThread(non_blocking_fn, 2)                                                                                                                                     
        reactor.callInThread(non_blocking_fn, 3)                                                                                                                                     
        reactor.callInThread(non_blocking_fn, 4)                                                                                                                                     
```

this will avoid the problematic code path and start all the threads as you would expect.

**Comment by twisted-trac** [comment]
|<img alt="bkambach's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|bkambach commented|
|-|-|

Glyph,

Thanks for the feedback.  If I understand comment #7 correctly, there is a difference with respect to the threadpool when callInThread is invoked *before*, versus *after*, the reactor is running?

WRT comment #5, the documentation refers to the default reactor having a minimum threadpool size of 5 (https://twistedmatrix.com/documents/current/core/howto/threading.html#managing-the-reactor-thread-pool):
>The default size of the thread pool depends on the reactor being used; the default reactor uses a minimum size of 5 and a maximum size of 10.
Am I interpreting this incorrectly?  I assumed it meant that 5 worker threads always exist to service jobs.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Replying to [bkambach](#note_8):
> Glyph,
> 
> Thanks for the feedback.  If I understand comment #7 correctly, there is a difference with respect to the threadpool when callInThread is invoked *before*, versus *after*, the reactor is running?

Yes; before the reactor is running the threadpool isn't started, and therefore won't actually schedule any work.  And there's a bug in the code that hands off work to threads from an idle threadpool when it starts; it only ever starts a single thread.  So this is definitely a bug.

> WRT comment #5, the documentation refers to the default reactor having a minimum threadpool size of 5 (https://twistedmatrix.com/documents/current/core/howto/threading.html#managing-the-reactor-thread-pool):
> >The default size of the thread pool depends on the reactor being used; the default reactor uses a minimum size of 5 and a maximum size of 10.
> Am I interpreting this incorrectly?  I assumed it meant that 5 worker threads always exist to service jobs.

Your interpretation is perfectly accurate, the problem is just that it's factually wrong :-).  The default pool has a minthreads of 0 and a max of 10: https://github.com/twisted/twisted/blob/b572e7e1ffa299f9751721405d5a5536995df85e/twisted/internet/base.py#L957

Would you mind opening a separate doc bug for this?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

(In [46147]) Branching to threadstarter-8090.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

Builders are going.  I think this squashes the bug decisively.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set owner to @glyph|
|-|-|

Thanks for the changes.

performAllCoordination needs a better docstring. The current one just expand the method name. I think that as a test helper, the docstring should describe why it exists and when/how/why it should be used.

for test_tooMuchWorkBeforeStarting, maybe we should create a self.threadpool with a custom max value so that we can assert based on that value... rather than on self.threadpool.max which is decided outside of the test.

-------

In MemoryPool instead of ```def _pool(pool, currentLimit, threadFactory):``` maybe we can have ```def _pool(this, currentLimit, threadFactory):`` ? Maybe we can move MemoryPool definition outside of setUp method.

-------

Maybe this is just a personal comment so feel free to ignore it.

Rather than having ```self.threadpool = MemoryPool(0, 10)``` in setUp, I prefer to have a ```threadpool = MemoryPool(0, 10)``` inside the test. In this way it is much easier to read/see that is the system under test and how it is initialized and what are its properties.

-----

The current tests look ok, but users will never run in production with a MemoryPool.

Is it hard to have the test as async one and do a real, functional (integration/system) test for the unmodified ThreadPool code?

I understand the role of unit tests, but if later we change someting in ThreadPool we might break this functionality and the current tests will still pass.

I assume there are many layered unit tests which should prevent such a case... but if is not hard to write/run functional code with unmodified code then I prefer to go for this route.

-------

Please address my comments and merge or resubmit.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="bkambach's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|bkambach commented|
|-|-|

Replying to [glyph](#note_9):
> 
> Your interpretation is perfectly accurate, the problem is just that it's factually wrong :-).  The default pool has a minthreads of 0 and a max of 10: https://github.com/twisted/twisted/blob/b572e7e1ffa299f9751721405d5a5536995df85e/twisted/internet/base.py#L957
> 
> Would you mind opening a separate doc bug for this?
I opened [#8104](https://github.com/twisted/twisted/issues/8104)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

Replying to [adiroiban](#note_12):
> Thanks for the changes.

Thanks for the review.

> performAllCoordination needs a better docstring. The current one just expand the method name. I think that as a test helper, the docstring should describe why it exists and when/how/why it should be used.

You're absolutely right, I added some detail.

> for test_tooMuchWorkBeforeStarting, maybe we should create a self.threadpool with a custom max value so that we can assert based on that value... rather than on self.threadpool.max which is decided outside of the test.

This value is specifically initialized by `setUp`, though.

> In MemoryPool instead of ```def _pool(pool, currentLimit, threadFactory):``` maybe we can have ```def _pool(this, currentLimit, threadFactory):`` ? Maybe we can move MemoryPool definition outside of setUp method.

I keep saying that we should extract text fixtures into separate test modules but then when the time comes to write a unit test I just go ahead and write a `setUp` again with some inner classes or whatever.  So I made this a proper helper and moved the threadpool subclass up to the top level.

> Maybe this is just a personal comment so feel free to ignore it.

I think you got at some useful stuff, so thank you :).

> Rather than having ```self.threadpool = MemoryPool(0, 10)``` in setUp, I prefer to have a ```threadpool = MemoryPool(0, 10)``` inside the test. In this way it is much easier to read/see that is the system under test and how it is initialized and what are its properties.

I made a `PoolHelper` that gets instantiated in each test and takes the arguments to a `ThreadPool`.

> The current tests look ok, but users will never run in production with a MemoryPool.

> Is it hard to have the test as async one and do a real, functional (integration/system) test for the unmodified ThreadPool code?

Yes, it's basically impossible to do that.  Particularly in this case: if we execute real code in threads, there is no moment where the thread count will reliably, deterministically be at a certain level.  Of course we could add `sleep()`s and such.

> I understand the role of unit tests, but if later we change someting in ThreadPool we might break this functionality and the current tests will still pass.

The problem with multithreading is that it's basically impossible to prove anything interesting about its properties by unit testing, or even by cursory integration testing.  Even if we could write something that would fail, it would fail with a hang or a crash.

> I assume there are many layered unit tests which should prevent such a case... but if is not hard to write/run functional code with unmodified code then I prefer to go for this route.

The entire architecture of `twisted._threads` is designed around making tests like these as meaningful as possible.  Of course they are not _perfect_ but they are also exercising _mostly_ the real implementation, stubbing out only the smallest bit necessary to make the tests deterministic.

> Please address my comments and merge or resubmit.

Since I only made test changes to improve comprehensibility I will go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph set status to `closed`|
|-|-|

(In [46250]) Merge threadstarter-8090: callInThread before run

Author: glyph

Reviewer: adiroiban

Fixes: [#8090](https://github.com/twisted/twisted/issues/8090)

twisted.python.threadpool.ThreadPool now properly starts enough threads to do
any work scheduled before ThreadPool.start() is called, such as when work is
scheduled in the reactor via reactor.callInThread() before reactor.run().
